controller/src/main/java/io/pravega/controller/server/rpc/grpc/v1/ControllerServiceImpl.java

```diff
@@ -343,23 +360,136 @@ public void deleteStream(StreamInfo request, StreamObserver<DeleteStreamStatus>
+                    authAndTokenPermission = AuthHandler.Permissions.READ_UPDATE;
+                }
+                this.grpcAuthHelper.checkAuthorization(resource, authAndTokenPermission);
+                return this.grpcAuthHelper.createDelegationToken(streamResource, authAndTokenPermission);
+            } else {
+                log.trace("Access operation was {} for request with scope {} and stream {}",
+                        translate(request.getAccessOperation()), request.getScope(), request.getStream());
+
+                // The resource string that'll be used in the delegation token for use of the segment store
+                final String tokenResource;
+
+                // The operation that'll be specified as granted for the resource in the token. The bearer of the token
+                // will be allowed to perform the specified operation.
+                final AuthHandler.Permissions tokenPermission;
+
+                // This is the permission that the client is requesting to be assigned on the delegation token.
+                AuthHandler.Permissions requestedPermissions = authParams.requestedPermission();
+
+                if (authParams.isStreamUserDefined()) {
+                    // The operation itself requires the caller to possess read permissions.
+                    AuthHandler.Permissions minimumPermissions = AuthHandler.Permissions.READ;
+
+                    if (requestedPermissions.equals(AuthHandler.Permissions.READ_UPDATE) ||
+                            requestedPermissions.equals(minimumPermissions)) {
+                        this.grpcAuthHelper.checkAuthorization(streamResource, requestedPermissions);
+                        tokenResource = streamResource;
+                        tokenPermission = requestedPermissions;
+                    } else {
+                        // The minimum permission that the user must have to be able to invoke this call. This
+                        // authorizes the operation.
+                        this.grpcAuthHelper.checkAuthorization(streamResource, minimumPermissions);
+
+                        // Here, we check whether the user is authorized for the requested access.
+                        this.grpcAuthHelper.checkAuthorization(streamResource, requestedPermissions);
+                        tokenResource = streamResource;
+                        tokenPermission = requestedPermissions;
+                    }
+                } else {
+                    final AuthHandler.Permissions authorizationPermission;
+                    if (requestedPermissions.equals(AuthHandler.Permissions.READ_UPDATE)) {
+                        // Clients have no business requesting for delegation tokens with write permissions for
+                        // mark streams. Clients are allowed to read from a mark stream, but aren't allowed to write to
+                        // it. (Only Controller writes to it.)
+                        if (authParams.isMarkStream()) {
```

Comment:
> ...Or are we checking it elsewhere?

We are checking for external/internal streams via this condition: 
https://github.com/pravega/pravega/blob/fe8f7f527db3221ecfd39899a7e3281d6ae3b922/controller/src/main/java/io/pravega/controller/server/rpc/grpc/v1/ControllerServiceImpl.java#L445

Accordingly, line 470 is hit only if the stream is an internal stream. 

> Also, if delegation tokens generated here are also used by the controller, then are we limiting access to mark streams by the controller?

The Controller obtains the delegation token via the `LocalController` (a client class), which doesn't make the gRPC call, but rather invokes a local call: 
https://github.com/pravega/pravega/blob/fe8f7f527db3221ecfd39899a7e3281d6ae3b922/controller/src/main/java/io/pravega/controller/server/eventProcessor/LocalController.java#L521

Reply:
Ok, I understand that the controller gets its delegation tokens elsewhere. On mark streams, given that we are saying that internal streams might be accessed by external clients for debug/administration, this is the behavior I'd expect:

1- If a client is requesting permission to `READ_UPDATE` a stream _S_, then it will not get automatically permission to `READ_UPDATE` the corresponding mark stream of _S_.
2- If a client is requesting `READ_UPDATE` access to a mark stream explicitly, say for debugging/administration purposes like with other internal streams, then it authorizes the access in the case the access configuration allows it. 

I'm not sure we are making a distinction between these two scenarios, and I'm trying to make sure that we are not blocking access to mark streams for say admin tools while access to other internal streams is authorized.

Reply:
> 1- If a client is requesting permission to READ_UPDATE a stream S, then it will not get automatically permission to READ_UPDATE the corresponding mark stream of S.

That is true. We only return a delegation token for the resource (stream, etc.) that the client has explicitly requested it for. 

Clients are now blocked by the server from getting a delegation token with `READ_UPDATE` for a mark stream. This means, if a client requests a token for a markstream with `READ_UPDATE`, they will encounter an error. 

> 2- If a client is requesting READ_UPDATE access to a mark stream explicitly, say for debugging/administration purposes like with other internal streams, then it authorizes the access in the case the access configuration allows it.

Clients are allowed to obtain a delegation token with `READ` on a mark stream, but not `READ_UPDATE` (which will be blocked). Since we definitively know clients have no business requesting for `READ_UPDATE` to a mark-stream, we'd rather block it right now and open it up later if and when the need arises.

Reply:
It is correct that regular clients have no business writing to a mark stream, I'm +1 for not grating `READ_UPDATE` access like that.

I'm concerned that we might not be doing it for internal streams consistently, which might be necessary for administration/recovery purposes. This is not a case of a regular user trying to append to/read from a stream, but an administrator trying to understand or fix a problem. In such cases, is it a possibility that the administrator might need to have `READ_UPDATE` access to one or more mark streams?

Reply:
In this PR, I’ve changed how delegation tokens are authorized for mark streams (they now require access on the corresponding streams), so based on a previous discussion that mark streams are not writable by clients, I’ve blocked writes to them. In the future if the need arises for clients to write to them, we can enable that with due consideration.

Reply:
I don’t want to make such a change for all internal streams, as that’s be a much broader change, and something that’s out of scope of this PR.

Reply:
Yes, we should first classify the internal streams into three categories - 
1. internal streams created for fulfilling a user function and users have full control over it (e.g. reader group streams).
2. internal stream created for fulfilling a user function but users dont have write access to it (e.g. mark stream).
3. internal streams that controller process creates for its own purposes. users have no business looking at or doing anything to do with these streams. 

First and second streams are already covered by these changes as a special case. 
For third, we should not prevent `READ_UPDATE` requests on authorization. this is because segment store also posts events into them. And admin tool may also require READ_UPDATE to allow them to post events to initiate some work to remedy some bad state in the system.

Reply:
Would it make sense to create an issue to define and implement an authorization model for internal streams?

Reply:
Yes, I will do so later.

Reply:
An issue to capture the open item we discussed last.

https://github.com/pravega/pravega/issues/5332